middleware/service/source/manager/handle.cpp

```diff
@@ -791,7 +791,20 @@ namespace casual
                         }
 
                         // check if some of the rest has _routes_
-                        for( auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
+                        // Note:
+                        // The following commented for stmt generates a warning about
+                        // use of a possibly uninitialized variable in
+                        // commmon/range.h when compiling with g++-11 (and 10)
+                        // without --debug. The warning did not occur with g++-9.
+                        // Introducing a local variable to hold the result of the
+                        // call to intersection() makes the warning go away.
+                        //
+                        // Unknown why the warning appears. Possibly some rule about
+                        // lifetimes, order of evaluation or similiar affects the
+                        // involved code. 
+                        //for( auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
+                        auto intersect = algorithm::sorted::intersection( message.content.services(), reply.content.services());
+                        for( auto& name : std::get< 1>( intersect))
```

Comment:
I would like you to try  `for( const auto& name : std::get< 1>( <same-as-the-original>)`

const ref, should extend the lifetime of  name in "whatever the compiler generates for the range-for-loop".

Reply:
Tried with following code:
```
                        for( const auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
                           if( auto found = algorithm::find( state.reverse_routes, name))
                              reply.routes.services.emplace_back( found->first, found->second);

                        local::optional::send( state, message.process.ipc, reply);


```
Same error:

compile: /export/home/guo/casual/middleware/service/source/manager/handle.cpp
error:
In file included from ../common/include/common/view/binary.h:10,
                 from ../common/include/common/serialize/traits.h:15,
                 from ../common/include/common/serialize/macro.h:10,
                 from ../common/include/common/strong/type.h:11,
                 from ../common/include/common/message/type.h:12,
                 from ../common/include/common/message/service.h:10,
                 from include/service/manager/state.h:9,
                 from include/service/manager/handle.h:12,
                 from /export/home/guo/casual/middleware/service/source/manager/handle.cpp:8:
In member function ‘constexpr casual::common::Range<Iter>::iterator casual::common::Range<Iter>::begin() const [with Iter = __gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >]’,
    inlined from ‘casual::service::manager::handle::local::{anonymous}::domain::discovery::internal::request(casual::service::manager::State&)::<lambda(casual::domain::message::discovery::internal::Request&)>’ at /export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:149:
../common/include/common/range.h:94:61: error: ‘<unnamed>.std::tuple<casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<0, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Head_base<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, false>::_M_head_impl.casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > > >::m_first.__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >::_M_current’ may be used uninitialized [-Werror=maybe-uninitialized]
   94 |          constexpr iterator begin() const noexcept { return m_first;}
      |                                                             ^~~~~~~
/export/home/guo/casual/middleware/service/source/manager/handle.cpp: In lambda function:
/export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:94: note: ‘<anonymous>’ declared here
  805 |                         for( const auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
      |                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../common/include/common/view/binary.h:10,
                 from ../common/include/common/serialize/traits.h:15,
                 from ../common/include/common/serialize/macro.h:10,
                 from ../common/include/common/strong/type.h:11,
                 from ../common/include/common/message/type.h:12,
                 from ../common/include/common/message/service.h:10,
                 from include/service/manager/state.h:9,
                 from include/service/manager/handle.h:12,
                 from /export/home/guo/casual/middleware/service/source/manager/handle.cpp:8:
In member function ‘constexpr casual::common::Range<Iter>::iterator casual::common::Range<Iter>::end() const [with Iter = __gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >]’,
    inlined from ‘casual::service::manager::handle::local::{anonymous}::domain::discovery::internal::request(casual::service::manager::State&)::<lambda(casual::domain::message::discovery::internal::Request&)>’ at /export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:149:
../common/include/common/range.h:95:59: error: ‘<unnamed>.std::tuple<casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<0, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Head_base<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, false>::_M_head_impl.casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > > >::m_last.__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >::_M_current’ may be used uninitialized [-Werror=maybe-uninitialized]
   95 |          constexpr iterator end() const noexcept { return m_last;}
      |                                                           ^~~~~~
/export/home/guo/casual/middleware/service/source/manager/handle.cpp: In lambda function:
/export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:94: note: ‘<anonymous>’ declared here
  805 |                         for( const auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
      |                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors

aborting due to errors
error building...
[

Reply:
Very long error message. Not so easy to see where/why the warning(error) occurs. Line 805 in the lambda mentioned at the end of the messages is the line "for( const aiuo& name: std::get< 1>(...", but the warnings occur in  casual::common::Range::begin() and  casual::common::Range::end(). 
The "real" problem may be "deeper down" than the for loop. If there IS a problem and not a compiler bug :-). Compiler bugs are not common, and as the error occurs with both gcc-10 and gcc-11 so the bug has been there for some time if there really is a problem. Makes a compiler bug less likely.
Guessing from the message there is some agressive inlining going on in the compiler...

Reply:
on the  "especially for debug-mode."...
This particular warning does NOT occur when compiling with --debug. It only occurs when --debug is NOT used. Likely because the compiler is not doing the same level of data flow analysis when compiling with debug. I noticed that compiling this file takes significant time, around 10 seconds or so, so the compiler does a significant amount of work :-)

Reply:
Ahh, I got it backwards :)

Reply:
Can you paste the error again within  _three ticks_ (markdown syntax to make it "termialish")

Reply:
```compile: /export/home/guo/casual/middleware/service/source/manager/handle.cpp
error:
In file included from ../common/include/common/view/binary.h:10,
                 from ../common/include/common/serialize/traits.h:15,
                 from ../common/include/common/serialize/macro.h:10,
                 from ../common/include/common/strong/type.h:11,
                 from ../common/include/common/message/type.h:12,
                 from ../common/include/common/message/service.h:10,
                 from include/service/manager/state.h:9,
                 from include/service/manager/handle.h:12,
                 from /export/home/guo/casual/middleware/service/source/manager/handle.cpp:8:
In member function ‘constexpr casual::common::Range<Iter>::iterator casual::common::Range<Iter>::begin() const [with Iter = __gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >]’,
    inlined from ‘casual::service::manager::handle::local::{anonymous}::domain::discovery::internal::request(casual::service::manager::State&)::<lambda(casual::domain::message::discovery::internal::Request&)>’ at /export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:149:
../common/include/common/range.h:94:61: error: ‘<unnamed>.std::tuple<casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<0, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Head_base<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, false>::_M_head_impl.casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > > >::m_first.__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >::_M_current’ may be used uninitialized [-Werror=maybe-uninitialized]
   94 |          constexpr iterator begin() const noexcept { return m_first;}
      |                                                             ^~~~~~~
/export/home/guo/casual/middleware/service/source/manager/handle.cpp: In lambda function:
/export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:94: note: ‘<anonymous>’ declared here
  805 |                         for( const auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
      |                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../common/include/common/view/binary.h:10,
                 from ../common/include/common/serialize/traits.h:15,
                 from ../common/include/common/serialize/macro.h:10,
                 from ../common/include/common/strong/type.h:11,
                 from ../common/include/common/message/type.h:12,
                 from ../common/include/common/message/service.h:10,
                 from include/service/manager/state.h:9,
                 from include/service/manager/handle.h:12,
                 from /export/home/guo/casual/middleware/service/source/manager/handle.cpp:8:
In member function ‘constexpr casual::common::Range<Iter>::iterator casual::common::Range<Iter>::end() const [with Iter = __gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >]’,
    inlined from ‘casual::service::manager::handle::local::{anonymous}::domain::discovery::internal::request(casual::service::manager::State&)::<lambda(casual::domain::message::discovery::internal::Request&)>’ at /export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:149:
../common/include/common/range.h:95:59: error: ‘<unnamed>.std::tuple<casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<0, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Head_base<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, false>::_M_head_impl.casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > > >::m_last.__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >::_M_current’ may be used uninitialized [-Werror=maybe-uninitialized]
   95 |          constexpr iterator end() const noexcept { return m_last;}
      |                                                           ^~~~~~
/export/home/guo/casual/middleware/service/source/manager/handle.cpp: In lambda function:
/export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:94: note: ‘<anonymous>’ declared here
  805 |                         for( const auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
      |                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors

aborting due to errors
error building...
[infra] guo@bvin01-unisb-01:~/casual $```

Reply:
That do make the message look better ;-)

Reply:
@lazan 
I saw the following in a list of new C++ 20 features [https://github.com/AnthonyCalandra/modern-cpp-features#range-based-for-loop-with-initializer](https://github.com/AnthonyCalandra/modern-cpp-features#range-based-for-loop-with-initializer) . Is this related to the warning I get with g++-11? I.e the "...elegant solution to a common lifetime problem". Introducing a local variable before the for is somewhat related in effect to what I assume this new language featuire does. 

Range-based for loop with initializer
This feature simplifies common code patterns, helps keep scopes tight, and offers an elegant solution to a common lifetime problem.

for (auto v = std::vector{1, 2, 3}; auto& e : v) {
  std::cout << e;
}
// prints "123"

Reply:
I have been looking at [https://en.cppreference.com/w/cpp/language/range-for](https://en.cppreference.com/w/cpp/language/range-for) for a while. In particular the section about "temporary range expression". 

"If range-expression returns a temporary, its lifetime is extended until the end of the loop, as indicated by binding to the forwarding reference __range. Lifetimes of all temporaries within range-expression are _not (until C++23)_ extended if they would otherwise be destroyed at the end of range-expression (since C++23)."

This problem may be worked around using init-statement:

for (T thing = foo(); auto& x : thing.items()) { /* ... */ } // OK (since C++20)"


The init statement solution is semantically very like the effect of the local variable I added. 
 
I am to tired to think through everything now (and I am not a real C++ specialist), but I get the feeling that lifetimes are at play here. Intersection() returns a tuple that is a temporary, and std::get< 1> also returns a temporary (does it?) based on item 1 in the tuple. Now it is possible that the code is "correct" but I get the feeling that the compiler for some reason or another thinks otherwise.

It does say "MAY be used uninitialized", If taken literally (!) this should mean that there exist a path where it may be used uninitialized, but it is not always uninitialized. 

This is the single place in all of casual that gets this warning, so something is special/unique with this code. 
Should probably make up small examples and try them out.